dom.go

```diff
@@ -709,29 +712,110 @@ func Text(text string, m ...MarkupOrChild) *HTML {
 }
 
 // Rerender causes the body of the given component (i.e. the HTML returned by
-// the Component's Render method) to be re-rendered.
+// the component's Render method) to be re-rendered.
 //
 // If the component has not been rendered before, Rerender panics. If the
 // component was previously unmounted, Rerender is no-op.
+//
+// Rerender operates efficiently by batching renders together. As a result,
+// there is no guarantee that a calls to Rerender will map 1:1 with calls to
+// the component's Render method. For example, two calls to Rerender may
+// result in only one call to the component's Render method.
 func Rerender(c Component) {
 	if c == nil {
 		panic("vecty: Rerender illegally called with a nil Component argument")
 	}
-	prevRender := c.Context().prevRender
-	if prevRender == nil {
-		panic("vecty: Rerender invoked on Component that has never been rendered")
-	}
 	if c.Context().unmounted {
 		return
 	}
-	nextRender, skip, pendingMounts := renderComponent(c, prevRender)
-	if skip {
+	batch.add(c)
+}
+
+// batchRenderer handles component re-renders by queueing and deduplicating
+// them, to be rendered on the next animation frame (via requestAnimationFrame).
+type batchRenderer struct {
+	// batch contains the list of pending components to render.
+	batch []Component
+	// idx maps components to batch indexes to allow dedup, retaining order.
+	idx map[Component]int
+	// scheduled tracks whether a batch has been scheduled for processing.
+	scheduled bool
+}
+
+// add a Component to the pending batch.
+func (b *batchRenderer) add(c Component) {
+	if i, ok := b.idx[c]; ok {
+		// Delete previously queued render.
+		copy(b.batch[i:], b.batch[i+1:])
+		b.batch[len(b.batch)-1] = nil
+		b.batch = b.batch[:len(b.batch)-1]
+	}
+	// Append and index component.
+	b.batch = append(b.batch, c)
+	b.idx[c] = len(b.batch) - 1
+	// If we're not already scheduled for a render batch, request a render on
+	// the next frame.
+	if !b.scheduled {
+		b.scheduled = true
+		requestAnimationFrame(b.render)
+	}
+}
+
+// render the pending batch.
+// TODO(pdf): Add tests for time budget and multi-pass renders.
+func (b *batchRenderer) render(startTime float64) {
+	// If the batch is empty, mark as unscheduled, and stop render cycle.
+	if len(b.batch) == 0 {
+		b.scheduled = false
 		return
 	}
-	// mountUnmount to check for node replacement
-	mountUnmount(nextRender, prevRender)
-	replaceNode(nextRender.node, prevRender.node)
-	mount(pendingMounts...)
+
+	// Drain the current batch.
+	pending := b.batch
+	b.batch = nil
+	b.idx = make(map[Component]int)
+
+	// Process batch.
+	for i, c := range pending {
+		// Skip unmounted components.
+		if c.Context().unmounted {
+			continue
+		}
+
+		// Check for remaining time budget, targeting 60fps (~16ms per frame).
+		if i > 0 {
```

Comment:
It would be more clear to me to have this logic at the end of the `for` loop instead of in an `i > 0` conditional.

Reply:
If we put it at the end we need a condition like `if len(pending) - i > 1`, and offset the avg denominator and indexes.  I'll move it if you'd prefer, but not sure it makes things a lot clearer.

Reply:
Oh, got it, I missed that. Leave as-is then. My 2 am brain is not very functional it seems :)

Reply:
I'm not 100% today either, so I appreciate the eyes as always :+1: